### PR TITLE
Implement feather printer and parser

### PR DESCRIPTION
**Pull request by balavinaithirthan**
1. We built the parser and printer plugin for the feather data format. It can be used similarly to other data formats in Tenzir, such as JSON and YAML. The new feather parser and printer allow parsing and printing through operators such as write, to, and from. The parser is implemented in an event-driven format. As input, the parser receives an upstream generator of chunk_ptr (byte representation), which returns a generator of table_slice for downstream operators. In the parser, a separate stream of record batches is created. Bytes from the generator, one group at a time, are converted into record batches and pushed into a record batch queue. The parser blocks until the bytes are entirely in the queue, when the parser retrieves the record batch and co_yields. Due to the synchronous nature of the parser, we are guaranteed in-order batching. Contrastingly, the printer is a blocking function with a generator of table slices as input and a generator of chunks as output. As a blocking function, the entire generator of table_slice input is written to a byte buffer before being co_yielded. 

6. Change log: The existing feather_store plugin has been removed and exchanged with the parser/printer dual. Feather can now be used like other existing formats, such as YAML and JSON. For example, the user can read in feather input from a file or previous operator's output. Similarly, the user can write Feather to a file by naming the file extension .feather. Feather inputs with non-TQL types currently have undefined behavior. 

Fixes https://github.com/tenzir/issues/issues/1555

### CODE REVIEW

File: libtenzir/builtins/formats/feather2.cpp [hunk 1]
```diff
@@ -0,0 +1,203 @@
...
+#include <arrow/table.h>
+#include <arrow/util/key_value_metadata.h>
+#include <caf/expected.hpp>
+
+#include <queue>
+
+namespace tenzir::plugins::feather2 {
+namespace {
+
+class CallbackListener : public arrow::ipc::Listener {
```
Comment (dominiklohmann) [inline libtenzir/builtins/formats/feather2.cpp:30]: Should be snake_case, i.e., `callback_listener`.

File: libtenzir/builtins/formats/feather2.cpp [hunk 2]
```diff
@@ -0,0 +1,203 @@
...
+public:
+  CallbackListener() = default;
+
+  arrow::Status OnRecordBatchDecoded(
+    std::shared_ptr<arrow::RecordBatch> record_batch) override {
+    record_batch_buffer.push(record_batch);
```
Comment (dominiklohmann) [inline libtenzir/builtins/formats/feather2.cpp:36]:
```suggestion
    record_batch_buffer.push(std::move(record_batch));
```

File: libtenzir/builtins/formats/feather2.cpp [hunk 3]
```diff
@@ -0,0 +1,203 @@
...
+    while (!listener->record_batch_buffer.empty()) {
+      auto batch = listener->record_batch_buffer.front();
+      listener->record_batch_buffer.pop();
+      auto validate_status = batch->Validate();
+      TENZIR_ASSERT(validate_status.ok(), validate_status.ToString().c_str());
+      co_yield table_slice(batch);
+    }
+
+    if (done) {
+      TENZIR_INFO("returning");
```
Comment (dominiklohmann) [inline libtenzir/builtins/formats/feather2.cpp:78]: This log message looks like a leftover from debugging.

File: libtenzir/builtins/formats/feather2.cpp [hunk 4]
```diff
@@ -0,0 +1,203 @@
...
+    }
+
+    while (!listener->record_batch_buffer.empty()) {
+      auto batch = listener->record_batch_buffer.front();
+      listener->record_batch_buffer.pop();
+      auto validate_status = batch->Validate();
+      TENZIR_ASSERT(validate_status.ok(), validate_status.ToString().c_str());
+      co_yield table_slice(batch);
+    }
+
```
Comment (dominiklohmann) [inline libtenzir/builtins/formats/feather2.cpp:76]: Please remove the unnecessary empty lines within this file.

File: libtenzir/builtins/formats/feather2.cpp [hunk 5]
```diff
@@ -0,0 +1,203 @@
...
+  -> generator<chunk_ptr> {
+  auto batch = to_record_batch(input);
+  auto validate_status = batch->Validate();
+  TENZIR_ASSERT(validate_status.ok(), validate_status.ToString().c_str());
+
+  auto stream_writer_status = stream_writer->WriteRecordBatch(*batch);
+  if (!stream_writer_status.ok()) {
+    diagnostic::error("failed to write a record batch to the stream")
+      .note("{}", stream_writer_status.ToString())
+      .emit(ctrl.diagnostics());
```
Comment (dominiklohmann) [inline libtenzir/builtins/formats/feather2.cpp:97]:
```suggestion
    diagnostic::error("failed to write a record batch to the stream")
      .note("{}", stream_writer_status.ToString())
      .emit(ctrl.diagnostics());
    co_return;
```

File: libtenzir/builtins/formats/feather2.cpp [hunk 6]
```diff
@@ -0,0 +1,203 @@
...
+  if (!finished_buffer_result.ok()) {
+    diagnostic::error("failed to close the stream and return the buffer")
+      .note("{}", finished_buffer_result.status().ToString())
+      .emit(ctrl.diagnostics());
+    co_return;
+  }
+  auto finished_buffer = finished_buffer_result.ValueOrDie();
+  ;
+  auto chunk = chunk::make(finished_buffer);
+  co_yield chunk;
```
Comment (dominiklohmann) [inline libtenzir/builtins/formats/feather2.cpp:109]:
```suggestion
  co_yield chunk::make(finished_buffer_result.MoveValueUnsafe());
```

File: libtenzir/builtins/formats/feather2.cpp [hunk 7]
```diff
@@ -0,0 +1,203 @@
...
+  auto reset_buffer_result
+    = sink->Reset(); // the buffer is reinit with newly allocated memory because
+                     // the API does not offer a Reset that just clears the
+                     // original data
```
Comment (dominiklohmann) [inline libtenzir/builtins/formats/feather2.cpp:113]: Please put the comment in a full sentence before the code that affects it; putting it after the code in the same line usually breaks code formatting with clang-format.

Thanks for the explanatory comment here, I think that really helps! 🙏

File: libtenzir/builtins/formats/feather2.cpp [hunk 8]
```diff
@@ -0,0 +1,203 @@
...
+  if (!reset_buffer_result.ok()) {
+    diagnostic::error("failed to reset buffer")
+      .note("{}", reset_buffer_result.ToString())
+      .emit(ctrl.diagnostics());
+  }
+}
+
+// questions: should be include logic to ensure entire record batch recorded ()
+
```
Comment (dominiklohmann) [inline libtenzir/builtins/formats/feather2.cpp:122]: Per our discussion in the office, this seems to be a leftover comment that is irrelevant and can be removed.

File: libtenzir/builtins/formats/feather2.cpp [hunk 9]
```diff
@@ -0,0 +1,203 @@
...
+    return f.object(x).fields();
+  }
+};
+
+class feather2_printer final : public plugin_printer {
+public:
+  feather2_printer() = default;
+
+  auto name() const -> std::string override {
+    // FIXME: Rename this and the file to just feather.
```
Comment (dominiklohmann) [inline libtenzir/builtins/formats/feather2.cpp:146]: Please do this to all filenames and code after rebasing onto main after #4072 is merged.

File: libtenzir/builtins/formats/feather2.cpp [hunk 10]
```diff
@@ -0,0 +1,203 @@
...
+    return "feather2";
+  }
+
+  auto instantiate([[maybe_unused]] type input_schema,
+                   operator_control_plane& ctrl) const
+    -> caf::expected<std::unique_ptr<printer_instance>> override {
+    auto sink = arrow::io::BufferOutputStream::Create().ValueOrDie();
+    const arrow::ipc::IpcWriteOptions& options
+      = arrow::ipc::IpcWriteOptions::Defaults();
```
Comment (dominiklohmann) [inline libtenzir/builtins/formats/feather2.cpp:155]: Please create a follow-up story for setting the `codec` and `min_space_savings`.

File: libtenzir/builtins/formats/feather2.cpp [hunk 11]
```diff
@@ -0,0 +1,203 @@
...
+    auto schema = input_schema.to_arrow_schema();
+    auto stream_writer_result
+      = arrow::ipc::MakeStreamWriter(sink, schema, options);
```
Comment (dominiklohmann) [inline libtenzir/builtins/formats/feather2.cpp:158]:
```suggestion
    auto stream_writer_result
      = arrow::ipc::MakeStreamWriter(sink, input_schema.to_arrow_schema(), options);
    if (!stream_writer_result.ok()) {
      return diagnostic::error(...).to_error();
    }
```

File: libtenzir/builtins/formats/feather2.cpp [hunk 12]
```diff
@@ -0,0 +1,203 @@
...
+    auto stream_writer = stream_writer_result.ValueOrDie();
```
Comment (dominiklohmann) [inline libtenzir/builtins/formats/feather2.cpp:159]:
```suggestion
    auto stream_writer = stream_writer_result.MoveValueUnsafe();
```

File: libtenzir/builtins/formats/feather2.cpp [hunk 13]
```diff
@@ -0,0 +1,203 @@
...
+        .emit(ctrl.diagnostics());
+      co_return;
+    }
+
+    while (!listener->record_batch_buffer.empty()) {
+      auto batch = listener->record_batch_buffer.front();
+      listener->record_batch_buffer.pop();
+      auto validate_status = batch->Validate();
+      TENZIR_ASSERT(validate_status.ok(), validate_status.ToString().c_str());
+      co_yield table_slice(batch);
```
Comment (dominiklohmann) [inline libtenzir/builtins/formats/feather2.cpp:74]: What if this fails?

The `table_slice` constructor from a record batch currently assumes that all data is compatible with Tenzir's data format. Can we change that to not crash but rather error if there's a mismatch?

File: tenzir/integration/tests/feather2.bats
Comment (dominiklohmann) [inline tenzir/integration/tests/feather2.bats:1]: As discussed in the office, these need to be rewritten to be meaningful.

File: libtenzir/builtins/formats/feather2.cpp (hunk 13 above)
Comment (dominiklohmann) [reply libtenzir/builtins/formats/feather2.cpp:70]: We created a follow-up issue: https://github.com/tenzir/issues/issues/1715

File: tenzir/integration/tests/feather2.bats
Comment (dominiklohmann) [reply tenzir/integration/tests/feather2.bats:1]: We paired on this earlier today in the office.

File: libtenzir/builtins/formats/feather2.cpp (hunk 9 above)
Comment (dominiklohmann) [reply libtenzir/builtins/formats/feather2.cpp]: Given that #4072 turned out to be a lot more complicated than we had originally anticipated I ended up pushing the remaining changes to this PR and closed the other one.

File: tenzir/integration/tests/feather.bats [hunk 14]
```diff
@@ -0,0 +1,43 @@
...
+  if ! python -c pyarrow; then
+    pip install pyarrow
+  fi
+
+  tenzir "from ${BATS_TENZIR_DATADIR}/inputs/zeek/conn.log.gz read zeek-tsv | batch 512 | write feather" |
+    check python "${BATS_TENZIR_MISCDIR}/scripts/print-arrow-batch-size.py"
+}
+
+@test "invalid format" {
+  check ! tenzir 'version | write json | read feather'
```
Comment (balavinaithirthan) [inline tenzir/integration/tests/feather.bats:27]: We should change this test to not use version since version produces different length outputs

**Review by dominiklohmann — APPROVED** [review]
We pair-reviewed this in depth.

**Pull request merged**